### ISSUE TITLE
trivial_cast false positive with trait objects in `map`

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#23737]
This code will generate a warning:

```rust
trait T { fn foo(&self) {} }              
impl T for i32 {}                         
                                          
fn foo(t: Option<i32>) -> Option<Box<T>> {
    t.map(|t| Box::new(t) as Box<T>)      
}                                         
                                          
fn main() {}                             
```

```
foo.rs:5:15: 5:36 warning: trivial cast: `Box<i32>` as `Box<T>`, #[warn(trivial_casts)] on by default
foo.rs:5     t.map(|t| Box::new(t) as Box<T>)
                       ^~~~~~~~~~~~~~~~~~~~~
```

but when the cast is removed:

```
foo.rs:5:5: 5:27 error: mismatched types:
 expected `core::option::Option<Box<T>>`,
    found `core::option::Option<Box<i32>>`
(expected trait T,
    found i32) [E0308]
foo.rs:5     t.map(|t| Box::new(t))
             ^~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

**Comment by alexcrichton** [comment]
cc @nrc

**Comment by Ryman** [comment]
This does work with an intermediate binding, seems annoying though since ascription hasn't landed yet, especially for libs with `deny(warnings)`.

E.g.
```
t.map(|t| { let t: Box<T> = Box::new(t); t })
```

**Comment by nrc** [comment]
This isn't a bug - a trivial cast is one which can be done with coercion, not necessarily one that is given by subtyping. I agree with @Ryman that this is a bit annoying until type ascription lands. But note that in most cases there is enough information available to type inference for coercion to work without ascription or an intermediate variable.
